source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -7,53 +7,168 @@ namespace Extensions {
 namespace UdpFilters {
 namespace UdpProxy {
 
+UdpProxyFilter::UdpProxyFilter(Network::UdpReadFilterCallbacks& callbacks,
+                               const UdpProxyFilterConfigSharedPtr& config)
+    : UdpListenerReadFilter(callbacks), config_(config),
+      cluster_update_callbacks_(
+          config->clusterManager().addThreadLocalClusterUpdateCallbacks(*this)) {
+  Upstream::ThreadLocalCluster* cluster = config->clusterManager().get(config->cluster());
+  if (cluster != nullptr) {
+    onClusterAddOrUpdate(*cluster);
+  }
+}
+
+void UdpProxyFilter::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (cluster.info()->name() != config_->cluster()) {
+    return;
+  }
+
+  ENVOY_LOG(debug, "udp proxy: attaching to cluster {}", cluster.info()->name());
+  cluster_info_.emplace(*this, cluster);
+}
+
+void UdpProxyFilter::onClusterRemoval(const std::string& cluster) {
+  if (cluster != config_->cluster()) {
+    return;
+  }
+
+  ENVOY_LOG(debug, "udp proxy: detaching from cluster {}", cluster);
+  cluster_info_.reset();
+}
+
 void UdpProxyFilter::onData(Network::UdpRecvData& data) {
+  if (!cluster_info_.has_value()) {
+    config_->stats().downstream_sess_no_route_.inc();
+    return;
+  }
+
+  cluster_info_.value().onData(data);
+}
+
+void UdpProxyFilter::onReceiveError(Api::IoError::IoErrorCode) {
+  config_->stats().downstream_sess_rx_errors_.inc();
+}
+
+UdpProxyFilter::ClusterInfo::ClusterInfo(UdpProxyFilter& filter,
+                                         Upstream::ThreadLocalCluster& cluster)
+    : filter_(filter), cluster_(cluster),
+      cluster_stats_(generateStats(cluster.info()->statsScope())),
+      member_update_cb_handle_(cluster.prioritySet().addMemberUpdateCb(
+          [this](const Upstream::HostVector&, const Upstream::HostVector& hosts_removed) {
+            for (const auto& host : hosts_removed) {
+              // This is similar to removeSession() but slightly different due to removeSession()
+              // also handling deletion of the host to session map entry if there are no sessions
+              // left. It would be nice to unify the logic but that can be cleaned up later.
+              auto host_sessions_it = host_to_sessions_.find(host.get());
+              if (host_sessions_it != host_to_sessions_.end()) {
+                for (const auto& session : host_sessions_it->second) {
+                  ASSERT(sessions_.count(session) == 1);
+                  sessions_.erase(session);
+                }
+                host_to_sessions_.erase(host_sessions_it);
+              }
```

Comment:
Sorry what do you mean? In this case the host has been removed from the cluster. On the next data, we will make a new session and pick a new host. Or do you mean to mark the session as having no host and then pick a new one in the data path in this case? If the latter, IMO this is a low frequency operation and it probably doesn't matter, but I'm happy to change it do doing that if you feel strongly about it.

Reply:
I was thinking whenever we remove a host, we actually don't need to remove the "session" associated with it but just change the session to send to a different host because after removal any incoming packet on that session can cause creation of another session with the same source/destination addresses but a different host anyway.  Same as unhealthy session below. 
But I can't tell how much tearing down and bringing up sessions can affect CPU when this happens. So probably it doesn't worth specialize this process.

Reply:
Yeah I agree it could be done more efficiently as you point out, but I don't think it's a super common operation and for now I would rather keep the logic as simple as possible. We can always optimize later.